src/types/resource/stream.rs

```diff
@@ -136,7 +136,38 @@ impl Stream {
     }
     pub fn streaming_url(&self, streaming_server_url: Option<&Url>) -> Option<String> {
         match (&self.source, streaming_server_url) {
-            (StreamSource::Url { url }, _) if url.scheme() != "magnet" => Some(url.to_string()),
+            (StreamSource::Url { url }, _) if url.scheme() != "magnet" => {
+                match (streaming_server_url, &self.behavior_hints.proxy_headers) {
+                    (
+                        Some(streaming_server_url),
+                        Some(StreamProxyHeaders { request, response }),
+                    ) => {
+                        let mut streaming_url = streaming_server_url.to_owned();
+                        let mut proxy_query = form_urlencoded::Serializer::new(String::new());
+                        let origin = format!("{}://{}", url.scheme(), url.authority());
+                        proxy_query.append_pair("d", origin.as_str());
+                        proxy_query.extend_pairs(
+                            request
+                                .iter()
+                                .map(|header| ("h", format!("{}:{}", header.0, header.1))),
+                        );
+                        proxy_query.extend_pairs(
+                            response
+                                .iter()
+                                .map(|header| ("r", format!("{}:{}", header.0, header.1))),
+                        );
+                        streaming_url
+                            .path_segments_mut()
+                            .ok()?
+                            .push("proxy")
+                            .push(proxy_query.finish().as_str())
+                            .push(&url.path()[1..]);
```

Comment:
It would be nice to document this format. I suppose this is a format that the Streaming server defines?
It could be beneficial to make a trait and implement it on Url for building this streaming url with proxy.
1. it's much easier to document
2. It will make it easy to test later on as well as a unit and not part of the `Stream::streaming_url`

Reply:
SInce this is used only on specific conditions and is needed here only, I personally wouldn't make it as a trait and leave it here, where all other streaming links are built.

Reply:
ok, I just wanted to isolate it instead of testing with a full Stream object.